utils/strip-metadata/src/lib.rs

```diff
@@ -0,0 +1,946 @@
+/// - Iterate over the type IDs again, mapping those found in the metadata to the new IDs that calling
+///   `self.types.retain(..)` handed back.
+pub trait StripMetadata {
+    /// Strip out any pallets and runtime APIs for which the provided filter functions return false.
+    fn strip_metadata<PalletFilter, RuntimeApiFilter>(
+        &mut self,
+        keep_pallet: PalletFilter,
+        keep_runtime_api: RuntimeApiFilter,
+    ) where
+        PalletFilter: Fn(&str) -> bool,
+        RuntimeApiFilter: Fn(&str) -> bool;
+}
+
+impl StripMetadata for v14::RuntimeMetadataV14 {
+    fn strip_metadata<PalletFilter, RuntimeApiFilter>(
+        &mut self,
+        keep_pallet: PalletFilter,
+        _keep_runtime_api: RuntimeApiFilter,
+    ) where
+        PalletFilter: Fn(&str) -> bool,
+        RuntimeApiFilter: Fn(&str) -> bool,
+    {
+        // Throw away pallets we don't care about:
+        self.pallets.retain(|pallet| keep_pallet(&pallet.name));
+
+        // Now, only retain types we care about in the registry:
+        retain_types(self);
+    }
+}
+
+impl StripMetadata for v15::RuntimeMetadataV15 {
+    fn strip_metadata<PalletFilter, RuntimeApiFilter>(
+        &mut self,
+        keep_pallet: PalletFilter,
+        keep_runtime_api: RuntimeApiFilter,
+    ) where
+        PalletFilter: Fn(&str) -> bool,
+        RuntimeApiFilter: Fn(&str) -> bool,
+    {
+        // Throw away pallets and runtime APIs we don't care about:
+        self.pallets.retain(|pallet| keep_pallet(&pallet.name));
+        self.apis.retain(|api| keep_runtime_api(&api.name));
+
+        // Now, only retain types we care about in the registry:
+        retain_types(self);
+    }
+}
+
+impl StripMetadata for v16::RuntimeMetadataV16 {
+    fn strip_metadata<PalletFilter, RuntimeApiFilter>(
+        &mut self,
+        keep_pallet: PalletFilter,
+        keep_runtime_api: RuntimeApiFilter,
+    ) where
+        PalletFilter: Fn(&str) -> bool,
+        RuntimeApiFilter: Fn(&str) -> bool,
+    {
+        // Throw away pallets and runtime APIs we don't care about:
+        self.pallets.retain(|pallet| keep_pallet(&pallet.name));
+        self.apis.retain(|api| keep_runtime_api(&api.name));
+
+        // Now, only retain types we care about in the registry:
+        retain_types(self);
+    }
+}
+
+fn retain_types<M: GetTypes + IterateTypeIds>(m: &mut M) {
+    // We want to preserve this type even if it's not used anywhere:
+    let dispatch_err_type_id = find_dispatch_error_type(m.get_types_mut());
+
+    // Iterate over the type IDs and retain any that we still need:
+    let keep_these_ids: BTreeSet<u32> = m
+        .iter_type_ids_mut()
+        .map(|id| *id)
+        .chain(Some(dispatch_err_type_id))
+        .collect();
+
+    let new_ids = m.get_types_mut().retain(|id| keep_these_ids.contains(&id));
+
+    // Map IDs found in the metadata to new ones as needed after the retaining:
+    m.iter_type_ids_mut().for_each(|id| {
```

Comment:
nit: maybe `filter_map().for_each()` ?

Reply:
I went the other way and simplified to a normal for loop:

```rust
for id in m.iter_type_ids_mut() {
    if let Some(new_id) = new_ids.get(id) {
        *id = *new_id;
    };
}
```

filter_map.for_each would have felt a bit ugly I think ie something like:

```rust
m.iter_type_ids_mut()
 .filter_map(|id| new_ids.get(id).map(|new_id| (id, new_id))
 .for_each(|(id, new_id)| *id = new_id)
```